contracts/javascore/xcall/src/main/java/foundation/icon/btp/xcall/CallServiceImpl.java

```diff
@@ -114,36 +101,38 @@ private void cleanupCallRequest(BigInteger sn) {
     public BigInteger sendCallMessage(String _to, byte[] _data, @Optional byte[] _rollback) {
         BigInteger sn = getNextSn();
         Address caller = Context.getCaller();
-        Context.require(caller.isContract() || _rollback == null, "RollbackNotPossible");
+        Context.require(caller.isContract(), "Only contract is allowed to send call message ");
         Context.require(_data.length <= MAX_DATA_SIZE, "MaxDataSizeExceeded");
         Context.require(_rollback == null || _rollback.length <= MAX_ROLLBACK_SIZE, "MaxRollbackSizeExceeded");
 
-        boolean needResponse = _rollback != null;
+        boolean needResponse = _rollback != null && _rollback.length > 0;
 
-        if (needResponse) {
-            CallRequest req = new CallRequest(caller, _to, _rollback);
-            requests.set(sn, req);
-        }
+        CallRequest req = new CallRequest(caller, _to, _rollback);
+        requests.set(sn, req);
 
         CSMessageRequest msgReq = new CSMessageRequest(caller.toString(), _to, sn, needResponse, _data);
 
         Packet pct = new Packet();
-        pct.setSequence(getNextSn());
-        pct.setData(msgReq.toString());
+        pct.setSequence(sn);
+        pct.setData(createMessage(CSMessage.REQUEST, msgReq.toBytes()));
         pct.setDestinationPort(getDestinationPort());
         pct.setDestinationChannel(getDestinationChannel());
         pct.setSourcePort(getSourcePort());
         pct.setSourceChannel(getSourceChannel());
 
+        BigInteger _timoutHeight = this.timeoutHeight.get();
         Height hgt = new Height();
-        hgt.setRevisionHeight(timeoutHeight.get());
-        hgt.setRevisionNumber(BigInteger.ZERO);
+        BigInteger timeoutHeight = BigInteger.valueOf(Context.getBlockHeight()).add(_timoutHeight);
+        hgt.setRevisionHeight(timeoutHeight);
+
         pct.setTimeoutHeight(hgt);
 
-        pct.setTimeoutTimestamp(BigInteger.ZERO);
+        //TODO set packet timeout
+        BigInteger timeout = BigInteger.valueOf(Context.getBlockTimestamp())
```

Comment:
Keep this as 0 instead which in IBC means can't be timed out by timestamp. Only height

Reply:
done